packages/account/src/providers/provider.test.ts

```diff
@@ -206,34 +208,22 @@ describe('Provider', () => {
   });
 
   it('can change the provider url of the current instance', async () => {
-    const providerUrl1 = FUEL_NETWORK_URL;
-    const providerUrl2 = 'http://127.0.0.1:8080/graphql';
+    using provider = await setupTestProvider();
 
-    const provider = await Provider.create(providerUrl1, {
-      fetch: (url: string, options: FetchRequestOptions) =>
-        getCustomFetch('getVersion', { nodeInfo: { nodeVersion: url } })(url, options),
-    });
-
-    expect(provider.url).toBe(providerUrl1);
-    expect(await provider.getVersion()).toEqual(providerUrl1);
-
-    const spyFetchChainAndNodeInfo = vi
-      .spyOn(Provider.prototype, 'fetchChainAndNodeInfo')
-      .mockResolvedValue({
-        chain: {} as ChainInfo,
-        nodeInfo: {} as NodeInfo,
-      });
+    const { cleanup, url } = await launchNode({ port: '0' });
```

Comment:
If we have automatic resource disposal, what is `cleanup` doing here? Might be good to make reference to the functions available off `launchNode` in the docs also.

Reply:
> If we have automatic resource disposal, what is cleanup doing here?

`launchNode` doesn't have automatic resource disposal, only `setupTestProvider` and above. 

This test is using `fetchChainAndNodeInfo` as a proxy for verifying that the URL did get changed and if I had used an additional `using provider2 = setupTestProvider()`, the `fetchChainAndNodeInfo` function would've been called on that other provider's creation for caching purposes and doing `await provider.connect(url)` would've skipped the call to `fetchChainAndNodeInfo`. That's why I had to explicitly use `launchNode`, so that another `Provider` isn't created.

> [..] Might be good to make reference to the functions available off `launchNode` in the docs also.

I didn't explain any of the other helpers besides `TestNodeLauncher` just so that the users aren't confused about various options that are all, in the end, covered by `TestNodeLauncher`. Everything that can be passed to `launchNode` can also be passed to `TestNodeLauncher`.

Reply:
Well, I'm unsure about the automatic disposal thing, it can be confusing.

Also, could we export `TestNodeLauncher` as a function named `launchNodeForTest(..)` instead?

I'd say it should be side-by-side with `launchNode(..)`, use common terminology, and have a similar API.

 - `launchNode.ts`
 - `launchNodeForTest.ts` _👈 new suggested name_

> _The latter as a wrapper around the former._

Reply:
> Well, I'm unsure about the automatic disposal thing, it can be confusing.

I'd say that if we make it clear to the users how it's used that they'll be happy with not having to call `cleanup` manually. It's a well-known concept across multiple languages, as was mentioned in the motivation for the feature in [TS](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#using-declarations-and-explicit-resource-management):
> You might be familiar with [using declarations in C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using), [with statements in Python](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement), or [try-with-resource declarations in Java](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html). These are all similar to JavaScript’s new using keyword, and provide a similar explicit way to perform a "tear-down" of an object at the end of a scope.

I left the option to opt out of it by passing `false` as the second `dispose` parameter and be able to call `cleanup` manually:
```ts
const { cleanup, contracts, wallets, provider } = await TestNodeLauncher.launch({}, false);
```

> [..] Also, could we export TestNodeLauncher as a function named launchNodeForTest(..) instead?

Sure, I can do it.

> [..] I'd say it should be side-by-side with launchNode(..), use common terminology, and have a similar API.

Can't put them side-by-side because `TestNodeLauncher`/`launchNodeForTest` needs `@fuel-ts/contract` which is unavailable in `@fuel-ts/account`. The terminology and APIs are similar already, one expanding on the other.

Reply:
> Cool, then we should consider having automatic disposal for both, with the same API.

That I can do.

> Perhaps both launchers could be inside `@fuel-ts/utils/test-utils`. 🙂

`launchNode` is used in the `fuels` CLI so it's not only a test util, but it can be moved into `@fuel-ts/utils`. As for `TestNodeLauncher`/`launchNodeForTest`, I can't move it anywhere except higher in the dependency chain because it needs `Contract` and `ContractFactory` to function.

Reply:
> [...] Also, could we export `TestNodeLauncher` as a function named `launchNodeForTest(..)` instead?

https://github.com/FuelLabs/fuels-ts/pull/1356/commits/d3fd97673c2ffd3538c247bde3bd2bc0e454a4bd

Reply:
Moving `launchNode` to `@fuel-ts/utils` currently isn't possible because it's got a dependency on `Signer`. We'd have to remove the logic around `process.env.GENESIS_SECRET` to make it happen. I'm actually a fan of that because this PR allows for ad hoc wallet generation, but it's IMO better to do it in a subsequent PR.

As for making `launchNode` disposable, after thinking about it a bit, I'm against it more than I am for it. It'd only be used in `setupTestProvider` and `autoStartFuelCore`, and in both places it'd be used as not disposable because of additional logic around it after the node is launched. After we merge this PR, it'll be removed from all tests as well, so all references will be only in the two aforementioned functions. IMO it'd be too much ceremony in the utility for no substantial gain.

Reply:
We would need `process.env.GENESIS_SECRET` to set the key for live environments, taken from the vault.

Reply:
> We would need process.env.GENESIS_SECRET to set the key for live environments, taken from the vault.

But that's something not related to `launchNode`. The live environment tests won't be using the helpers this PR introduces.